src/core/reactor.cc

```diff
@@ -1921,6 +1925,191 @@ reactor::inotify_add_watch(int fd, std::string_view path, uint32_t flags) {
     });
 }
 
+future<std::tuple<pid_t, int, int, int>>
+reactor::spawn(std::string_view pathname,
+               std::vector<sstring> argv,
+               std::vector<sstring> env) {
+    return do_with(pid_t{},
+                   std::array<int, 2>{-1, -1},
+                   std::array<int, 2>{-1, -1},
+                   std::array<int, 2>{-1, -1},
+                   sstring(pathname),
+                   posix_spawn_file_actions_t{},
+                   posix_spawnattr_t{},
+                   std::move(argv),
+                   std::move(env),
+                   [this](auto& child_pid, auto& cin_pipe, auto& cout_pipe, auto& cerr_pipe, auto& pathname, auto& actions, auto& attr, auto& argv, auto& env) {
+        auto create_cin_pipe = _thread_pool->submit<syscall_result<int>>([&cin_pipe] {
+            return wrap_syscall<int>(::pipe2(cin_pipe.data(), O_NONBLOCK));
+        });
+        auto create_cout_pipe = _thread_pool->submit<syscall_result<int>>([&cout_pipe] {
+            return wrap_syscall<int>(::pipe2(cout_pipe.data(), O_NONBLOCK));
+        });
+        auto create_cerr_pipe = _thread_pool->submit<syscall_result<int>>([&cerr_pipe] {
+            return wrap_syscall<int>(::pipe2(cerr_pipe.data(), O_NONBLOCK));
+        });
+        static constexpr int pipefd_read_end = 0;
+        static constexpr int pipefd_write_end = 1;
+        return when_all_succeed(std::move(create_cin_pipe),
+                                std::move(create_cout_pipe),
+                                std::move(create_cerr_pipe)).then_unpack([&cin_pipe, &cout_pipe, &cerr_pipe] (syscall_result<int> cin_ret,
+                                                                                                              syscall_result<int> cout_ret,
+                                                                                                              syscall_result<int> cerr_ret) {
+            if (cin_ret.result == 0 && cout_ret.result == 0 && cerr_ret.result == 0) {
+                return;
+            }
+            for (auto pipe : {cin_pipe, cout_pipe, cerr_pipe}) {
+                if (pipe[0] >= 0) {
+                    ::close(pipe[pipefd_read_end]);
+                    ::close(pipe[pipefd_write_end]);
+                }
+            }
+            for (auto& ret : {cin_ret, cout_ret, cerr_ret}) {
+                ret.throw_if_error();
+            }
+        }).then([&child_pid, &cin_pipe, &cout_pipe, &cerr_pipe, &pathname, &actions, &attr, &argv, &env, this] {
+            // the args and envs parameters passed to posix_spawn() should be array of pointers, and
+            // last one should be a null pointer.
+            std::vector<const char*> argvp;
+            std::transform(argv.cbegin(), argv.cend(), std::back_inserter(argvp),
+                           [](auto& s) { return s.c_str(); });
+            argvp.push_back(nullptr);
```

Comment:
What if it vector::push_back() throws ENOMEM, will pipe descriptors leak?

Reply:
will address this one in the next revision.

Reply:
fixed.